### PR TITLE
Adding service to set custom particles

### PR DESCRIPTION
**Pull request by vik748**
This commit adds a service called "amcl/set_particles" to the amcl node to accept a Pose array of particles to set custom particles in AMCL.

### CODE REVIEW

File: amcl/src/amcl_node.cpp [hunk 1]
```diff
@@ -1020,6 +1028,97 @@ AmclNode::globalLocalizationCallback(std_srvs::Empty::Request& req,
...
+
+
+bool AmclNode::setParticlesCallback(amcl::amcl_particles::Request& req,
+                                    amcl::amcl_particles::Response& res)
+{  
+  int no_of_particles =  req.pose_array_msg.poses.size();
+  ROS_INFO("Received set particles srv call, Pose Array Header seq = %i",req.pose_array_msg.header.seq);
+  ROS_INFO("Received set particles srv call, Pose Array Header stamp = %0.4f",req.pose_array_msg.header.stamp.toSec());
+  if (no_of_particles == max_particles_)
+	{
```
Comment (mikeferguson) [inline amcl/src/amcl_node.cpp]: indentation is off here -- looks like you've got some tabs -- indent should 2 spaces

File: amcl/src/amcl_node.cpp [hunk 2]
```diff
@@ -1020,6 +1028,97 @@ AmclNode::globalLocalizationCallback(std_srvs::Empty::Request& req,
...
+  return p;
+}
+
+
+bool AmclNode::setParticlesCallback(amcl::amcl_particles::Request& req,
+                                    amcl::amcl_particles::Response& res)
+{  
+  int no_of_particles =  req.pose_array_msg.poses.size();
+  ROS_INFO("Received set particles srv call, Pose Array Header seq = %i",req.pose_array_msg.header.seq);
+  ROS_INFO("Received set particles srv call, Pose Array Header stamp = %0.4f",req.pose_array_msg.header.stamp.toSec());
```
Comment (mikeferguson) [inline amcl/src/amcl_node.cpp]: All this output feels a little spammy (7 ROS_INFOs per call). I'd suggest converting all but one of them to ROS_DEBUG_NAMED() calls -- with a good namespace so it's easy to turn on.

File: amcl/src/amcl_node.cpp [hunk 3]
```diff
@@ -1020,6 +1028,97 @@ AmclNode::globalLocalizationCallback(std_srvs::Empty::Request& req,
...
+pf_vector_t AmclNode::customPoseGenerator(void* arg)
+{
+  geometry_msgs::PoseArray* parray = (geometry_msgs::PoseArray*) arg;
+  geometry_msgs::Pose le = parray->poses.back();
+  tfScalar roll, pitch, yaw ;
+  
+  tf::Quaternion quat(le.orientation.x,le.orientation.y,le.orientation.z,le.orientation.w);
+  tf::Matrix3x3(quat).getRPY(roll, pitch, yaw);
+  
+  //std::cout<<"Size of array "<<parray->poses.size()<<std::endl;
```
Comment (mikeferguson) [inline amcl/src/amcl_node.cpp]: please convert to ROS_DEBUG_NAMED, or remove

**Review by mikeferguson — CHANGES_REQUESTED** [review]
A couple of style issues noted.

A separate issue: it is a little late to add such a new feature to Kinetic. Also, we can't just cherry pick this into Melodic (since it has been converted to use TF2). If possibly, I'd really prefer we update this to use TF2 and merge it into Melodic only. Thoughts?

**Comment by ghost** [comment]
@mikeferguson I have updated the node to address your formatting comments.  As for whether to include it into Kinetic, I'll let you decide the norm.  I just wanted to leave this here so that anyone interested could have access to it.
As for the update for Melodic, I haven't really installed / used Melodic yet.  So I won't get around to updating this to TF2 anytime soon.
Just to understand the process, if / when I get to committing those changes in Melodic.  Do I just update the melodic branch in my fork or would that be a completely new pull request?

**Comment by mikeferguson** [comment]
Targets a dead release (kinetic). If this is still of interest, please open a PR to noetic-devel branch

**Pull request closed without merging**